### ISSUE TITLE
OSError instead of ImportError when libusb binary is missing

### ISSUE THREAD

**Issue by matejcik** [vpelletier/python-libusb1#37]
`_loadLibrary` is called at import time. It tries to `dlopen()` the right binary, performs some sort of fallback if the call fails with `OSError`, but if the library can't be found, the original `OSError` is raised.
(see: 
https://github.com/vpelletier/python-libusb1/blob/eb5c0333d2a2e5af3eadad32f907abdcc934876b/usb1/libusb1.py#L196 )

That means that the following correct-looking code ends up wrong:
```python
try:
	import usb1
except ImportError:
	print("libusb not available")
	usb1 = None
```

the `raise` in question should probably be changed to:
```python
	raise ImportError("could not find libusb binary") from oserr
```
(with `except OSError as oserr` at start of the block)

**Comment by vpelletier** [comment]
What would be the issue with the following code ?
```python
usb1 = None
try:
    import usb1
except ImportError:
    print("python-libusb1 not available")
except OSError:
    print("Native libusb1 not available")
```

**Comment by matejcik** [comment]
Iiiiiii.... don't think an `import` statement is ever supposed to raise anything other than a subclass of `ImportError`, unless circumstances are particularly severe?

The Python docs state the following about `ImportError`:
_Raised when the import statement has troubles trying to load a module._
Which is the case here.

IMHO it would be appropriate to raise an `OSError`, if you had something like `usb1.load_library()` call, and that would fail when the binary is not found.
The issue here is that the error is raised _in an import statement_. From my point of view as a programmer, _I'm not doing anything_ yet.
(I mean, in the same vein, you could argue that "`usb1/__init__.py ` does not exist" is _technically_ an `OSError`, as it results from some syscall failing with `ENOTFOUND`. And yet...)

`try ... except ImportError` is the idiomatic way to guard against failing `import` statements. By raising something else, you're pushing knowledge and responsibility to users of the library - now I need to know that libusb, unlike the majority of other libraries, has two different things that can go wrong when importing. And that I need to code against both of them separately.

**Comment by vpelletier** [comment]
> Which is the case here.

My reading is that this describes how python itself raises such error (when import statement on its own cannot succeed) rather than prescribing how to use this exception class.

> *I'm not doing anything yet.*

The alternative would be to postpone linker invocation: do not look for libusb1 at import time, but when USBContext is instanciated for the first time for example.

Which is an idea which already somewhat surfaced when I've been asked how to customise library lookup path. My answer at that time was to just do the same as one would do with any non-python program: use whatever linker configuration knobs available on considered platform (`LD_LIBRARY_PATH` and equivalents). This would also break backward compatibility (as error source would move from import time to somewhere later), but this feels like it brings something (errors can be handled in a more convenient place than around import statements) instead of just removing a distinction of why `python-libusb1` cannot work.

> has two different things that can go wrong when importing. And that I need to code against both of them separately.

My intention was precisely that it made possible to tell your users about two different solutions for two different problems, instead of stuffing them into one umbrella exception.

While python3 improved things a lot with `raise ... from ...`, `ImportError` can be very misleading in my experience: maybe the module being import is not there, or maybe it is another module imported inside that is not, or maybe it is an import cycle and everything is actually present, or maybe it is some python code actively raising it for whatever other reason.

Then again, handling errors at import time is not the most convenient place, as it then encourages importing elsewhere than module toplevel beginning (I'm thinking of GUI programs, which would need to display an error dialog telling users to install libusb1).

I'll keep thinking on this to find a better way, any API idea to postpone and customise library lookup is welcome.

**Comment by vpelletier** [comment]
This is a tough one to crack.

- I intend to split `usb1.libusb1` into 2 pieces: constants, data structures and translated macros on one side, and a library loader & symbol extractor on the other side (not sure what "side" means exactly: a class/function within `usb1.libusb1`, or a separate submodule...). This will break compatibility with programs which import `usb1.libusb1` (and its old location, top-level module `libusb1`). They should really not, but... Legacy.
- `usb1.libusb1` internally works around an ABI incompatibility in an old-ish FreeBSD implementation of libusb1 API by introspecting the list of symbols exported by the library when on that platform. So structures (at least that one) cannot be defined until after the library has been loaded.
- It seems unlikely that loading two different versions of libusb1 in the same process would work, so some singleton enforcement will be needed.

An completely different (and simpler, and backward compatible) way could be to create a new exception which would inherit from both `OSError` (for backward-compatibility) and `ImportError` (to be less surprising) and raise that when the C library cannot be loaded. Not fully satisfying, but maybe a reasonable middle-ground.

**Comment by jayvdb** [comment]
> An completely different (and simpler, and backward compatible) way could be to create a new exception which would inherit from both OSError (for backward-compatibility) and ImportError (to be less surprising) and raise that when the C library cannot be loaded. 

Creating new exceptions would be ideal IMO.  Then the newer API can expose different exceptions for various failure conditions, with various other inheritance as appropriate (e.g. OSError), and as long as they all subclass `ImportError`, it will be backwards compatible with consumers expecting an `ImportError`.

**Comment by mcuee** [comment]
>While python3 improved things a lot with raise ... from ..., ImportError can be very misleading in my experience: maybe the module being import is not there, or maybe it is another module imported inside that is not, or maybe it is an import cycle and everything is actually present, or maybe it is some python code actively raising it for whatever other reason.

To me the documentation fix is enough to let the user know what to do (eg: to put libusb-1.0 library in the right locations). No need to over-engineering.

**Comment by matejcik** [comment]
> To me the documentation fix is enough to let the user know what to do (eg: to put libusb-1.0 library in the right locations).

This assumes that the end-user (a) sees the message and (b) has any idea whatsoever what "module", "libusb", or "full path" means.

This is a developer-side issue: if I want to code defensively around libusb not being available, I need to learn about the fact that it can raise an unconventional exception _at import time_ -- something that very well might never happen on a properly-configured (i.e. mine) computer.

**Comment by mcuee** [comment]
End user do not use python-libusb1 directly. If a developer choose to use python-libusb1 for his application, then he needs to bundle the necessary library, or at least tell the user how to do that under different OS.

**Comment by vpelletier** [comment]
I think I have a not-quite-scandalously-ugly solution:

- usb1.libusb1 stops loading libusb1 on import, but instead offers a `loadLibrary(libusb=None)` function which loads the library and creates all the module globals which need libusb to be loaded in order to be defined.
- backward-compatibility: usb1.libusb1 gets renamed, and a place-holder is put in its place
- usb1.USBContext gets a new `loadLibrary(libusb=None)` static method which calls the libusb1-level function. Its `__init__` calls this method without argument so the library get loaded when the context is first used.

This means:

- no more unusual exceptions being raised during module import. Instead, with unmodified application source, these happen during `USBContext` creation.
- applications which wish to customize the C library lookup can implement their own, and call `USBContext.loadLibrary` with whatever library object they loaded.

The downsides I see are:

- pylint is going to be mad at me for stuffing `globals()` dynamically. But I'm already doing it all over the place, so... meh ?
- applications which *did* handle the unusual exceptions at import time are going to break at `USBContext` construction time now.

Does this seem usable ?

Are the downsides acceptable ?

Are there other downsides that I did not think about ?

For more details, please see the [`wip` branch](/vpelletier/python-libusb1/branches/wip).

**Comment by vpelletier** [comment]
> The downsides I see are:

One more:

A maybe-not-obvious limitation is that the loading is possibly only once per loading of the module (...and as typically a module is not unloaded for the duration of a python process, this means once per process).

This avoids a lot of complexity, while this also means that it may not be possible to pick two unrelated projects using `python-libusb1` each on its own side and use them in the same process (in details: only one may customize the C library lookup).

**Comment by vpelletier** [comment]
@matejcik I am adding `libusb_wrap_sys_device` to `usb1.libusb1`, and it brings the question of another kind of import-time error detection: its second argument type is `intptr_t`, which does not exist in `ctypes`. The definition is that it can hold a pointer, so I can compare `sizeof(c_void_p)` with `sizeof(c_int32)` and `sizeof(c_int64)` and put whichever matches into a new `c_intptr_t` global.

What should I do if neither match ?

- raising `ImportError` feels wrong to me.
- raising `ValueError` or some other generic-ish exception type feels semantically better to me (in the very narrow context of "I cannot find a solution which satisfies the target value", ignoring the fact that we are at the module global level), but it is going against the desire of this bug report of avoiding anything other than `ImportError` during import.
- not raising anything means that `NameError` will be raised a few lines later when `c_intptr_t` is referenced. This feels worse than both of the above, as the exception source cannot give any useful insight, and overall this gives the feeling of a very bad module quality.

What would be your expectation in this case ?

Maybe the difference is the likelihood of the exception: non-powers-of-eight, <=16bits and >=128bits pointers all seem quite unlikely to be encountered by an end user first, while I agree that a missing/broken libusb-1.0 is a lot more common (...on archs which do not force me to bundle the library to get any decent experience, anyway...).

**Comment by matejcik** [comment]
I would not expect this to raise at import-time at all, tbh, but only when I try to call the problematic function.

(an open question is: would libusb itself work on a platform where intptr_t is something weird?)

It's also OK by me to declare that the import is all-or-nothing and that not having a matching size for intptr_t renders the library unusable as a whole -- in which case I'd find `ImportError` to be a perfectly appropriate way of raising "libusb can't be used".

**Comment by mcuee** [comment]
> (an open question is: would libusb itself work on a platform where intptr_t is something weird?)

Good question. So no need to worry about that. :-)

**Comment by mcuee** [comment]
https://gist.github.com/isislovecruft/836111acdccfbfb4a284
```
    import ctypes

    sizeof_voidp = ctypes.sizeof(ctypes.c_voidp)
    sizeof_int = ctypes.sizeof(ctypes.c_int)
    sizeof_long = ctypes.sizeof(ctypes.c_long)
    # Py_ssize_t (for ob_refcount) can be defined as having the size of
    # ssize_t (if our compiler and variant of C9* have ssize_t), otherwise
    # it has the size of "an integer large enough to store a pointer",
    # a.k.a. C99 intptr_t, but only if the sizeof(intptr_t) is equal to
    # the sizeof(void*).  Since we can't check the size of intptr_t from
    # ctypes, we'll just fallback to sizeof(void*) and hope for the best.
    # What could possibly go wrong?
    sizeof_Py_ssize_t =  ctypes.sizeof(
        ctypes.c_ssize_t if hasattr(ctypes, 'c_ssize_t') else ctypes.c_voidp)
```

**Comment by mcuee** [comment]
I can see some changes in the wip branch. I think the following output seems to be good.

Just a test without libusb-1.0.dll in the right directory. `import usb1` does not give errors which is correct.

```
(py39venv) C:\work\libusb\python-libusb1_wip [wip ≡]> python
Python 3.9.6 (tags/v3.9.6:db3ff76, Jun 28 2021, 15:26:21) [MSC v.1929 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import usb1
>>> with usb1.USBContext() as context:
...     for device in context.getDeviceIterator(skip_on_error=True):
...         print('ID %04x:%04x' % (device.getVendorID(), device.getProductID()), '->'.join(str(x) for x in ['Bus %03i' % (device.getBusNumber(), )] + device.getPortNumberList()), 'Device', device.getDeviceAddress())
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\work\libusb\python-libusb1_wip\usb1\__init__.py", line 2370, in __init__
    self.loadLibrary()
  File "C:\work\libusb\python-libusb1_wip\usb1\__init__.py", line 2364, in loadLibrary
    libusb1.loadLibrary(libusb=libusb)
  File "C:\work\libusb\python-libusb1_wip\usb1\_libusb1.py", line 201, in loadLibrary
    __loadLibrary(libusb)
  File "C:\work\libusb\python-libusb1_wip\usb1\_libusb1.py", line 217, in __loadLibrary
    libusb = __getLibrary()
  File "C:\work\libusb\python-libusb1_wip\usb1\_libusb1.py", line 158, in __getLibrary
    return dll_loader(filename, use_errno=True, use_last_error=True)
  File "C:\Python39\lib\ctypes\__init__.py", line 374, in __init__
    self._handle = _dlopen(self._name, mode)
FileNotFoundError: Could not find module 'libusb-1.0.dll' (or one of its dependencies). Try using the full path with constructor syntax.
>>>
```

**Comment by vpelletier** [comment]
> I can see some changes in the wip branch. I think the following output seems to be good (no change from the main branch).

Thanks for testing.

I'm still wondering whether the implicit `loadLibrary` call should happen during `USBContext.__init__` or `USBContext.open`: while the former is just fine given that `loadLibrary` is a static method, this may be just unusual enough that a caller may not realize they need to call that method on the class before instanciating it:

```
usb1.USBContext.loadLibrary(...)
with usb1.USBContext() as context:
  # do stuff
```

while anyone should naturally come up with a pattern like:

```python
context = usb1.USBContext() # would not raise here if I deffer implicit loadLibrary call until "open"
context.loadLibrary(...)
with context: # may raise here instead
  # do stuff
```

If I wait for `USBContext.open`, the only difference is that `self.__libusb_handle_events` will not be set until then, but the only method using it needs an open context anyway. And both patterns would remain applicable.

So it all depends on how much I should put the "this is not a normal method, it's affecting the class - and in fact the whole module - and not just an instance" in the mind of the developer.

**Comment by vpelletier** [comment]
> I'm still wondering whether the implicit `loadLibrary` call should happen during `USBContext.__init__` or `USBContext.open`

I resolved this by moving the static method into a module function. The intent should be clearer.

I also changed the code so `loadLibrary` returns whether the requested library was loaded rather than raise when it is not the case: caller code should not have to rely on a specific version of libusb-1.0, but rather *may* provide a version in case it knows where to locate it. And should it really want a specific version for some reason, then it can still check the return value.

I pushed these changes in `master` and dropped the `wip` branch.
